target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh

```diff
@@ -33,6 +33,20 @@ EOF
 
 platform_do_upgrade() {
 	case "$(board_name)" in
+	linksys,mr7350)
+		boot_part="$(fw_printenv -n boot_part)"
+		if [ "$boot_part" -eq "1" ]; then
+			fw_setenv boot_part 2
+			CI_KERNPART="alt_kernel"
+			CI_UBIPART="alt_rootfs"
+		else
+			fw_setenv boot_part 1
+			CI_UBIPART="rootfs"
+		fi
+		fw_setenv boot_part_ready 3
+		fw_setenv auto_recovery yes
```

Comment:
This is overkill.

```
		if [ "$(fw_printenv -n boot_part_ready)" -ne "3" ] ; then
			echo "boot_part_ready 3" >> $setenv_script
		fi
```
The entire NAND block will need to be rewritten. Whether an individual variable is modified or not does not matter. So this form adds a lot of code with no real benefit.

Reply:
I do not have this issue in my sysupgrade branch [1]. Is the OEM UBI volume removed so that there is enough space for `rootfs_data?

[1] https://github.com/mrnuke/openwrt/blob/2d2febc896bd61ed3371db70b53566f06ab5bc38/target/linux/qualcommax/ipq60xx/base-files/lib/upgrade/platform.sh#L130

Reply:
Do you have single `rootfs_data` partition?
For other Linksys devices we have two separate `rootfs_data` partitions: https://forum.openwrt.org/t/adding-openwrt-support-for-linksys-mx5300/92667/25

Reply:
Nope. One `rootfs_data` per UBI volume.

What is the output of  `ubinfo -a /dev/ubi0` ?

Reply:
> What is the output of `ubinfo -a /dev/ubi0` ?

```
ubi0
Volumes count:                           2
Logical eraseblock size:                 126976 bytes, 124.0 KiB
Total amount of logical eraseblocks:     592 (75169792 bytes, 71.6 MiB)
Amount of available logical eraseblocks: 0 (0 bytes)
Maximum count of volumes                 128
Count of bad physical eraseblocks:       0
Count of reserved physical eraseblocks:  40
Current maximum erase counter value:     3
Minimum input/output unit size:          2048 bytes
Character device major/minor:            245:0
Present volumes:                         0, 1

Volume ID:   0 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        326 LEBs (41394176 bytes, 39.4 MiB)
State:       OK
Name:        rootfs
Character device major/minor: 245:1
-----------------------------------
Volume ID:   1 (on ubi0)
Type:        dynamic
Alignment:   1
Size:        222 LEBs (28188672 bytes, 26.8 MiB)
State:       OK
Name:        rootfs_data
Character device major/minor: 245:2
```

Reply:
That looks as expected. Thank you!

Reply:
> Nope. One `rootfs_data` per UBI volume.

But you have two `rootfs` partitions:
- `rootfs` - mtd15
- `alt_rootfs` - mtd17
And after sysupgrade you are switched between mtd15 and mtd17.
Each `rootfs` partition contains only one `rootfs_data` volume.

Reply:
@Vladdrako can you show dmesg output how the partitions are created?
`Creating 20 MTD partitions on "qcom_nand.0":`

Reply:
Yes, that is exactly what I would expect.

Why do you believe that having separate `rootfs_data` on separate UBI devices is a problem, and why do you believe this would prevent config from being saved on sysupgrade?

Reply:
> can you show dmesg output how the partitions are created?

```
[    1.415891] 0x000000000000-0x000000180000 : "0:sbl1"
[    1.423027] 0x000000180000-0x000000280000 : "0:mibib"
[    1.427497] 0x000000280000-0x000000600000 : "0:qsee"
[    1.434178] 0x000000600000-0x000000680000 : "0:devcfg"
[    1.437105] 0x000000680000-0x000000700000 : "0:rpm"
[    1.442004] 0x000000700000-0x000000780000 : "0:cdt"
[    1.446832] 0x000000780000-0x000000800000 : "0:appsblenv"
[    1.451729] 0x000000800000-0x000000980000 : "0:appsbl"
[    1.457942] 0x000000980000-0x000000a00000 : "0:art"
[    1.462285] 0x000000a00000-0x000001300000 : "0:wififw"
[    1.473076] 0x000001300000-0x000001380000 : "0:ethphyfw"
[    1.473937] 0x000001380000-0x000001400000 : "u_env"
[    1.478274] 0x000001400000-0x000001440000 : "s_env"
[    1.482701] 0x000001440000-0x000001480000 : "devinfo"
[    1.487597] 0x000001480000-0x000001c80000 : "kernel"
[    1.498296] 0x000001c80000-0x000006680000 : "rootfs"
[    1.551163] mtd: setting mtd15 (rootfs) as root device
[    1.551458] mtdsplit: no squashfs found in "rootfs"
[    1.555244] 0x000006680000-0x000006e80000 : "alt_kernel"
[    1.566141] 0x000006e80000-0x00000b880000 : "alt_rootfs"
[    1.620480] 0x00000b880000-0x00000ba80000 : "sysdiag"
[    1.622480] 0x00000ba80000-0x00000fe80000 : "syscfg"
```

Reply:
> ```
> [    1.487597] 0x000001480000-0x000001c80000 : "kernel"
> [    1.498296] 0x000001c80000-0x000006680000 : "rootfs"
> [    1.555244] 0x000006680000-0x000006e80000 : "alt_kernel"
> [    1.566141] 0x000006e80000-0x00000b880000 : "alt_rootfs"
> ```

There are some differences here. In OEM firmware, the `kernel` partition overlaps the `rootfs` partition:
```
[    1.155578] 0x000001480000-0x000006680000 : "kernel"
[    1.218653] 0x000001c80000-0x000006680000 : "rootfs"
[    1.276153] 0x000006680000-0x00000b880000 : "alt_kernel"
[    1.339189] 0x000006e80000-0x00000b880000 : "alt_rootfs"
```
Keeping this layout will help in reverting to OEM firmware.

Reply:
> Yes, that is exactly what I would expect.
> 
> Why do you believe that having separate `rootfs_data` on separate UBI devices is a problem, and why do you believe this would prevent config from being saved on sysupgrade?

So every single file (even one created manually) is copied to another `rootfs_data` volume on the second `rootfs` partition during sysupgrade?

Reply:
My expectation is that an upgrade copies all relevant configuration to the `rootfs_data` volume of the UBI device given in `CI_UBIPART`. If that isn't happening,let me know what you observe, and I can help troubleshoot.

Note that this doesn't mean "every single file". sysupgrade shouldn't copy everything blindly.

Reply:
You are supposed to have `rootfs_data` per each rootfs UBI volume, that is normal.